crypto/rand/rand_vxworks.c

```diff
@@ -0,0 +1,164 @@
+#  endif
+# endif
+
+# if defined(RAND_SEED_VXRANDLIB)    
+#  include <randomNumGen.h>
+# endif
+
+/* Macro to convert two thirty two bit values into a sixty four bit one */
+# define TWO32TO64(a, b) ((((uint64_t)(a)) << 32) + (b))
+
+static uint64_t get_time_stamp(void)
+{
+    struct timespec ts;
+
+    if (clock_gettime(CLOCK_REALTIME, &ts) == 0)
+        return TWO32TO64(ts.tv_sec, ts.tv_nsec);
+    return time(NULL);
+}
+
+static uint64_t get_timer_bits(void)
+{
+    uint64_t res = OPENSSL_rdtsc();
+
+    if (res != 0)
+        return res;
+
+    {
+        struct timespec ts;
+
+        if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0)
+            return TWO32TO64(ts.tv_sec, ts.tv_nsec);
+    }
+    return time(NULL);
+}
+
+/* 
+ *	empty implementation 
+ *  vxworks does not need to init/cleanup or keep open the random lib
+ */
+int rand_pool_init(void)
+{
+    return 1;
+}
+
+void rand_pool_cleanup(void)
+{
+}
+
+void rand_pool_keep_random_devices_open(int keep)
+{
+}
+
+int rand_pool_add_additional_data(RAND_POOL *pool)
+{
+    struct {
+        CRYPTO_THREAD_ID tid;
+        uint64_t time;
+    } data = { 0 };
+
+    /*
+     * Add some noise from the thread id and a high resolution timer.
+     * The thread id adds a little randomness if the drbg is accessed
+     * concurrently (which is the case for the <master> drbg).
+     */
+    data.tid = CRYPTO_THREAD_get_current_id();
+    data.time = get_timer_bits();
+
+    return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
+}
+
+int rand_pool_add_nonce_data(RAND_POOL *pool)
+{
+    struct {
+        pid_t pid;
+        CRYPTO_THREAD_ID tid;
+        uint64_t time;
+    } data = { 0 };
+
+    /*
+     * Add process id, thread id, and a high resolution timestamp to
+     * ensure that the nonce is unique whith high probability for
+     * different process instances.
+     */
+    data.pid = getpid();
+    data.tid = CRYPTO_THREAD_get_current_id();
+
+    data.time = get_time_stamp();
+
+    return rand_pool_add(pool, (unsigned char *)&data, sizeof(data), 0);
+}
+
+size_t rand_pool_acquire_entropy(RAND_POOL *pool)
+{
+# if defined(RAND_SEED_VXRANDLIB)
+/* vxRandLib based entropy method */
+    size_t bytes_needed;
+    size_t entropy_available = 0;
+    unsigned char *buffer;
+
+    bytes_needed = rand_pool_bytes_needed(pool, 1 /*entropy_factor*/);
+    if (bytes_needed > 0) 
+    {
+        int retryCount = 0;
+        STATUS result = ERROR;
+        buffer = rand_pool_add_begin(pool, bytes_needed);
+        while ((result != OK) && (retryCount < 10))
+        {
+            result = randBytes(buffer, bytes_needed);
+            if (result == OK) 
+            {
+                rand_pool_add_end(pool, bytes_needed, 8 * bytes_needed);
+                entropy_available = rand_pool_entropy_available(pool);
+            } else 
+            {
+                /* give a minimum delay here to allow OS to collect more entropy */
+                taskDelay (1);

```

Comment:
is this a millisecond?

Reply:
it depends on the timer tick. in this case i think it is ok to wait "1 tick" whatever that means, this gives us a bit more time on slower systems

Reply:
BTW, no space between function name, and "("

Reply:
indeed i have (not sure if it was 1 or ~3 ticks though). the sw-random num lib generates entropy on each interrupt, the taskDelay at least has 1 timer interrupt happening, hopefuly more.

Reply:
Compare this to the linux add_interrupt_randomness, where it takes
64 interrupts and one second to acquire 1 bit of randomness !!!!
I need to say that I deeply mistrust this random source.

Reply:
based on my experiements (see below) i gave it a bit more of a wait, however the initial entropy gathered should be more than enough for seeding.